csharp/src/Google.Protobuf/JsonFormatter.cs

```diff
@@ -780,6 +783,62 @@ private static void HexEncodeUtf16CodeUnit(TextWriter writer, char c)
             writer.Write(Hex[(c >> 4) & 0xf]);
             writer.Write(Hex[(c >> 0) & 0xf]);
         }
+        
+        
+        private void WriteJsonObjectOpen(TextWriter writer, char openChar = '{')
+        {
+            writer.Write(openChar);
+            if (!settings.Pretty)
+            {
+                writer.Write(' ');
+            }
+        }
+
+        private void WriteJsonObjectClose(TextWriter writer, bool hasFields, int depth, char closeChar = '}')
+        {
+            if (hasFields)
+            {
+                if (settings.Pretty)
+                {
+                    writer.WriteLine();
+                    WriteIndentation(writer, depth);
+                }
+                else
+                {
+                    writer.Write(" ");
+                }
+            }
+            
+            writer.Write(closeChar);
+        }
+
+        private void WriteJsonObjectEntry(TextWriter writer, bool first, int depth)
+        {
+            if (settings.Pretty)
+            {
+                if (first)
+                {
+                    writer.WriteLine();
+                }
+                else
+                {
+                    writer.Write(MultilinePropertySeparator);
+                }
+                WriteIndentation(writer, depth);
+            }
+            else if (!first)
+            {
+                writer.Write(PropertySeparator);
+            }
+        }
+
+        private void WriteIndentation(TextWriter writer, int depth)
+        {
+            for (int i = 0; i < depth; i++)
+            {
+                writer.Write(settings.Indent);
+            }
+        }
```

Comment:
I've tried several ways to optimize this, for example by pre-calculating indentation for several levels, to reduce calls to `Write`. However, it added 10-20 lines of hard to read code, with a speed improvement of only 2-5%, so I think it's not worth it. If it's necessary, this can be optimized.

Although to be fair, optimization here would only benefit users of the new feature, and not improve the original implementation.

Reply:
I think it's absolutely fine as is, really.